packages/twenty-server/src/workspace/messaging/message/message.service.ts

```diff
@@ -81,4 +93,204 @@ export class MessageService {
+          await this.messageChannelMessageAssociationService.countByMessageExternalIdsAndMessageChannelId(
+            [message.externalId],
+            gmailMessageChannelId,
+            workspaceId,
+            manager,
+          );
+
+        if (existingMessageChannelMessageAssociationsCount > 0) {
+          return;
+        }
+
+        const savedOrExistingMessageThreadId =
+          await this.messageThreadService.saveMessageThreadOrReturnExistingMessageThread(
+            message.messageThreadExternalId,
+            dataSourceMetadata,
+            workspaceId,
+            manager,
+          );
+
+        const savedOrExistingMessageId =
+          await this.saveMessageOrReturnExistingMessage(
+            message,
+            savedOrExistingMessageThreadId,
+            connectedAccount,
+            dataSourceMetadata,
+            workspaceId,
+            manager,
+          );
+
+        await manager.query(
+          `INSERT INTO ${dataSourceMetadata.schema}."messageChannelMessageAssociation" ("messageChannelId", "messageId", "messageExternalId", "messageThreadId", "messageThreadExternalId") VALUES ($1, $2, $3, $4, $5)`,
+          [
+            gmailMessageChannelId,
+            savedOrExistingMessageId,
+            message.externalId,
+            savedOrExistingMessageThreadId,
+            message.messageThreadExternalId,
+          ],
+        );
+      });
+    }
+  }
+
+  private async saveMessageOrReturnExistingMessage(
+    message: GmailMessage,
+    messageThreadId: string,
+    connectedAccount: ObjectRecord<ConnectedAccountObjectMetadata>,
+    dataSourceMetadata: DataSourceEntity,
+    workspaceId: string,
+    manager: EntityManager,
+  ): Promise<string> {
+    const existingMessage = await this.messageService.getFirstByHeaderMessageId(
+      message.headerMessageId,
+      workspaceId,
+    );
+    const existingMessageId = existingMessage?.id;
+
+    if (existingMessageId) {
+      return Promise.resolve(existingMessageId);
+    }
+
+    const newMessageId = v4();
+
+    const messageDirection =
+      connectedAccount.handle === message.fromHandle ? 'outgoing' : 'incoming';
+
+    const receivedAt = new Date(parseInt(message.internalDate));
+
+    await manager.query(
+      `INSERT INTO ${dataSourceMetadata.schema}."message" ("id", "headerMessageId", "subject", "receivedAt", "direction", "messageThreadId", "text", "html") VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
+      [
+        newMessageId,
+        message.headerMessageId,
+        message.subject,
+        receivedAt,
+        messageDirection,
+        messageThreadId,
+        message.text,
+        message.html,
+      ],
+    );
+
+    await this.messageParticipantService.saveMessageParticipants(
+      message.participants,
+      newMessageId,
+      dataSourceMetadata,
+      manager,
+    );
+
+    return Promise.resolve(newMessageId);
+  }
+
+  public createQueriesFromMessageIds(
```

Comment:
This can be moved to its own util file `create-get-gmail-messages-queries-from-message-ids.util.ts` or something like that